text/0000-array-private-classes

```diff
@@ -0,0 +1,285 @@
+and does not try to access the iterator directly but is interested by the iterated
+items. When client code refers to the iterator type, that's generally useless and
+the code can be rewritten to be made shorter and more future proof.
+
+A [search on Github Pony code](https://github.com/search?q=%22ArrayValues%22+language%3APony&type=code)
+finds 24 files using the class `ArrayValues`, of which 6 are copies of `array.pony` file.
+
+For instance, in
+[xml2xpath.pony](https://github.com/redvers/pony-libxml2/blob/bbca5d98d48854bfec2c6ee110220873ecc4df34/pony-libxml2/xml2xpath.pony#L41),
+the code can be changed from
+
+```pony
+  fun values(): ArrayValues[Xml2node, this->Array[Xml2node]]^ ? =>
+    if (allocated) then
+      ArrayValues[Xml2node, this->Array[Xml2node]](nodearray)
+    else
+      error
+    end
+```
+
+to
+
+```pony
+  fun values(): RewindableIterator[Xml2node]^ ? =>
+    if (allocated) then
+      nodearray.values()
+    else
+      error
+    end
+```
+
+In this sample, the developer was not really concerned by the type of the iterator
+but that the `values` function must return an `RewindableIterator` over `Xml2node`.
+The new version makes the code simpler to understand.
+
+This change in `array.pony` and other collections will break such code but it
+can be easily adapted to use the new API. And it will make the standard library
+easier to learn by reducing the number of public types.
+
+# How We Test This
+
+Pony tests must continue to pass. No additional tests are need as after review
+the existing coverage in Pony standard library tests is sufficient
+
+# Drawbacks
+
+Will break any existing code that uses any of the classes that are currently
+public and will be made private by this RFC.
+
+# Alternatives
+
+Stay as is. Continue the
+[discussion on Zulip](https://ponylang.zulipchat.com/#narrow/stream/189959-RFCs/topic/Make.20Array.20iterators.20private).
+
+Instead of defining a new type of iterator with the `RewindableIterator` interface,
+we can consider only the rewindable part of it in an `Rewindable` interface:
+
+```pony
+interface Rewindable[A]
+  fun ref rewind(): Iterator[A]^
+    """
+    Start the iterator over again from the beginning.
+    """
+```
+
+Then one can create a rewindable iterator by composing these two interfaces:
+
+```pony
+type RewindableIterator[A] is (Iterator[A] & Rewindable[A])
+```
+
+The rewindable type can be used with other types than `Iterator`, like a data
+structure that would implement a rewindable property. This alternative was
+[put aside](https://github.com/ponylang/rfcs/pull/193#discussion_r780793165) to
+prevent name colisions in `builtin` with user-named types.
+
+# Unresolved questions
+
+Pony language definition is not precise about
+[type variance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science))
+and its impact in collections classes, particularly for covariant return types
+of functions (in this RFC, `keys`, `values` and `pairs`).
```

Comment:
Standard notions of covariance and contravariance do apply as normal in Pony, including covariant return types.

Do you still have open/unresolved questions on this topic? If so, I can probably help resolve them. Feel free to expand your questions here or hit me up in the Zulip chat to discuss further.

Once you're comfortable with the answers we can remove this from the "unresolved questions" section, since I don't think there are any Pony surprises related to this particular topic that are relevant to this RFC.